### ISSUE TITLE
`react/no-unused-prop-types` false positives when passing props as an argument

### ISSUE THREAD

**Issue by barakyosi** [yannickcr/eslint-plugin-react#2155]
In this scenario `something` is marked as unused although it is used.

```
class Hello extends Component {
  constructor (props) {
    super(props);
    this.doSomething(this.props);
  }
  doSomething(props) {
    if (props.something) {
      return true;
    }
  }
}
Hello.propTypes = {
  something: PropTypes.bool
}
```

It's similar to [#855](https://github.com/yannickcr/eslint-plugin-react/issues/855), but here the code is part of the class.

**Comment by ljharb** [comment]
This isn't really something we can statically analyze - in general, you shouldn't be passing around the entire props (or state) object. If you destructure `props` in the constructor, and pass `something` in explicitly, it should work fine.

**Comment by barakyosi** [comment]
I thought about detecting method calls with `props` argument, then look inside the method body and check usage.
Started to work on it but wasn't sure how to get the method body from the call expression.
Would be happy to know whether this check can be done (and if not what's the limitation)

**Comment by ljharb** [comment]
In your specific simple case, it could certainly be done - but, anyone from outside the file could call `doSomething` with an object that's *not* the props - so it's simply impossible to statically know it for all circumstances, only for a subset of usage patterns.

I don't think it's worth the complexity to try to target only *some* of the patterns, when the entire thing can be avoided by not having the antipattern of passing the props object around.

**Comment by barakyosi** [comment]
Sometimes passing all props makes it easier to follow. Especially when you need more than one prop and need to call it in different lifecycle methods.
I understand it's impossible to catch all scenarios. Would love to see a solution covering this one :)

**Comment by ljharb** [comment]
I'm sure it's subjective, but I would argue that in that case, you'd want to do this:
```
const { a, b, c } = props;
doSomething({ a, b, c });
```
You can still pass an object, but it's not the actual props object, it's an explicit object you created.

**Comment by barakyosi** [comment]
I agree your solution is better for this scenario, but if you need it in more than one place (in constructor & CWRP), or use many props inside `doSomething` - when adding a new prop you might forget to add it everywhere.

**Comment by ljharb** [comment]
I see that as a feature, not a bug, since if you do forget, your tests will fail :-)

**Comment by a-x-** [comment]
I created but-repro-repo with similar React.memo case: https://github.com/a-x-/eslint-memo-props-check

**Comment by a-x-** [comment]
In my case props passed as closure w/o function arguments

**Comment by ljharb** [comment]
This seems answered - don't pass the props or state objects anywhere.
